File: Aula_11_setembro/src/Package/Principal.java
package Package;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class Principal {
    public static void main(String[] args) {
        
        List<Integer> listaBolha = new ArrayList<>();
        List<Integer> listaNormal = new ArrayList<>();
        List<Integer> listaSelecao = new ArrayList<>();
        List<Integer> listaInsercao = new ArrayList<>();
        List<Integer> listaAgitacao = new ArrayList<>();
        List<Integer> listaPente = new ArrayList<>();
        List<Integer> listaShellsort = new ArrayList<>();
        List<Integer> listaQuicksort = new ArrayList<>();

        Util.popularListaAleatoria(listaBolha, 10000, 100, 20000);
        Util.popularListaAleatoria(listaNormal, 10000, 100, 20000);
        Util.popularListaAleatoria(listaSelecao, 10000, 100, 20000);
        Util.popularListaAleatoria(listaInsercao, 10000, 100, 20000);
        Util.popularListaAleatoria(listaAgitacao, 10000, 100, 20000);
        Util.popularListaAleatoria(listaPente, 10000, 100, 20000);
        Util.popularListaAleatoria(listaShellsort, 10000, 100, 20000);
        Util.popularListaAleatoria(listaQuicksort, 10000, 100, 20000);

        // Sort nativo
        long tempoInicio = System.currentTimeMillis();
        Collections.sort(listaNormal);
        long tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por sort nativo: " + (tempoFim - tempoInicio) + " ms");

        System.out.println();

        // Bolha
        tempoInicio = System.currentTimeMillis();
        int[] resultadoBolha = Ordenacao.bolha(listaBolha);
        tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por bolha: " + (tempoFim - tempoInicio) + " ms");
        System.out.println("Comparacoes: " + resultadoBolha[0]);
        System.out.println("Trocas: " + resultadoBolha[1]);

        System.out.println();

        // Selecao
        tempoInicio = System.currentTimeMillis();
        int[] resultadoSelecao = Ordenacao.selecao(listaSelecao);
        tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por selecao: " + (tempoFim - tempoInicio) + " ms");
        System.out.println("Comparacoes: " + resultadoSelecao[0]);
        System.out.println("Trocas: " + resultadoSelecao[1]);

        System.out.println();

        // Insercao
        tempoInicio = System.currentTimeMillis();
        int[] resultadoInsercao = Ordenacao.insercao(listaInsercao);
        tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por insercao: " + (tempoFim - tempoInicio) + " ms");
        System.out.println("Comparacoes: " + resultadoInsercao[0]);
        System.out.println("Trocas: " + resultadoInsercao[1]);
        
        System.out.println();


        /*
    * AGITAÇÃO - Shake Sort ou Cocktail

    - É baseado no Bolha, ou seja, é uma tentantiva de melhorar o Bolha.
    - Estável
    - Memória Interna
    - A estrutura possui 3 partes/porções (1a é a ordenada pelos menores; 2a é a desordenada; 3a é a ordenada pelos maiores)

    De fato é a aplicação do bolha da esquerda para direita e depois da direita para esquerda

    Há os índices/ponteiros ini e fim que marcam o início e o final da estrutura

    Observação: 
        i) se a estrutura estiver ordenada ou quase, o agitação terá excelente desempenho, como o bolha
        ii) atenção a lista deve ser duplamente encadeada  
    */
        // Agitacao:
        tempoInicio = System.currentTimeMillis();
        int[] resultadoAgitacao = Ordenacao.agitacao(listaAgitacao);
        tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por agitacao: " + (tempoFim - tempoInicio) + " ms");
        System.out.println("Comparacoes: " + resultadoAgitacao[0]);
        System.out.println("Trocas: " + resultadoAgitacao[1]);

        System.out.println();

        /*
         * PENTE - Comb Sort
    - É baseado no Bolha, ou seja, é uma tentantiva de melhorar o Bolha.

    ATENÇÃO: a partir deste método, há COMPARAÇÕES a uma distância X. Isso gera uma pré-organização da estrutura, diminuindo número de comparações e trocas.

    Instável
    Memória Interna
    A estrutura possui 2 porções/partes (1a é a ordenada; 2a é a desordenada)

    Há as variáveis clássicas do bolha: i, houveTroca, tmp.
    Há a variável distancia que é calculada pelo tamanho da estrutura dividido por 1.3

    Adequado somente para listas ou estruturas prontas tipo lista de uma linguagem de programação
         */

        // Pente (Comb Sort):
        tempoInicio = System.currentTimeMillis();
        int[] resultadoPente = Ordenacao.agitacao(listaPente);
        tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por Pente (Comb Sort): " + (tempoFim - tempoInicio) + " ms");
        System.out.println("Comparacoes: " + resultadoPente[0]);
        System.out.println("Trocas: " + resultadoPente[1]);

        System.out.println();

        /*
         * SHELLSORT 
        É baseado no Inserção, ou seja, é uma tentativa de melhoria via o uso da ANÁLISE A DISTÂNCIA (tipo pente) Instável 
        Memória Interna

        Não é adequado para listas encadeadas

    n = 7
    0   1   2   3   4   5   6   
    7   1   4   2   3   9   8       distancia = 4
         */

        // Shellsort:
        tempoInicio = System.currentTimeMillis();
        int[] resultadoShellsort = Ordenacao.shell(listaShellsort);
        tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por Shellsort: " + (tempoFim - tempoInicio) + " ms");
        System.out.println("Comparacoes: " + resultadoPente[0]);
        System.out.println("Trocas: " + resultadoPente[1]);

        System.out.println();

        //Quicksort:
        tempoInicio = System.currentTimeMillis();
        int[] resultadoQuickSort = Ordenacao.quickSort(listaQuicksort, 0, listaQuicksort.size() - 1);
        tempoFim = System.currentTimeMillis();
        System.out.println("Tempo da rotina ordenar por QuickSort: " + (tempoFim - tempoInicio) + " ms");
        System.out.println("Comparacoes: " + resultadoQuickSort[0]);
        System.out.println("Trocas: " + resultadoQuickSort[1]);
    }

}

File: Aula_25_Setembro_Python/Principal.py
from util import Util
from Ordenacao import Ordenacao as ord
import time
 
lista_bolha = []
lista_normal = []
lista_selecao = []
lista_insercao = []
lista_agitacao = []
lista_pente = []
lista_shell = []
lista_quick = []
lista_heap = []

Util.popular_lista_aleatoria(lista_bolha, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_normal, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_selecao, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_insercao, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_agitacao, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_pente, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_shell, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_quick, 10000, 100, 20000)
Util.popular_lista_aleatoria(lista_heap, 10000, 100, 20000)

#sort:
tempoInicio = time.time()
lista_normal.sort()
tempoFim = time.time()
print("Tempo da rotina ordenar por sort nativo: ", (tempoFim - tempoInicio) , "s", '\n')        

#Bolha: 
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.bolha(lista_bolha)
tempoFim = time.time()
print("Tempo da rotina ordenar por bolha: ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')  

#Selecao:
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.selecao(lista_selecao)
tempoFim = time.time()
print("Tempo da rotina ordenar por selecao: ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')  

#Insercao:
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.insercao(lista_insercao)
tempoFim = time.time()
print("Tempo da rotina ordenar por insercao: ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')

#Agitacao:
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.agitacao(lista_agitacao)
tempoFim = time.time()
print("Tempo da rotina ordenar por Agitacao: ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')

# Pente:
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.pente(lista_pente)
tempoFim = time.time()
print("Tempo da rotina ordenar por Pente (Comb Sort): ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')

#Shell:
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.shell(lista_shell)
tempoFim = time.time()
print("Tempo da rotina ordenar por Shellsort: ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')

#Quicksort:
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.quicksort(lista_quick, 0, len(lista_quick) - 1)
tempoFim = time.time()
print("Tempo da rotina ordenar por Quicksort: ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')

#Heapsort:
tempoInicio = time.time()
qtd_comparacoes, qtd_trocas = ord.heapsort(lista_heap)
tempoFim = time.time()
print("Tempo da rotina ordenar por Heapsort: ", (tempoFim - tempoInicio) , "s")      
print('Comparacoes:', qtd_comparacoes)
print('Trocas:', qtd_trocas, '\n')
